server/src/main/java/org/elasticsearch/cluster/metadata/FieldInferenceMetadata.java

```diff
@@ -0,0 +1,227 @@
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Function;
+
+/**
+ * Contains field inference information. This is necessary to add to cluster state as inference can be calculated in the coordinator
+ * node, which not necessarily has mapping information.
+ */
+public class FieldInferenceMetadata implements Diffable<FieldInferenceMetadata>, ToXContentFragment {
+
+    private final ImmutableOpenMap<String, FieldInference> fieldInferenceMap;
+
+    // Contains a lazily cached, reversed map of inferenceId -> fields
+    private volatile Map<String, Set<String>> fieldsForInferenceIds;
+
+    public static final FieldInferenceMetadata EMPTY = new FieldInferenceMetadata(ImmutableOpenMap.of());
+
+    public FieldInferenceMetadata(MappingLookup mappingLookup) {
+        ImmutableOpenMap.Builder<String, FieldInference> builder = ImmutableOpenMap.builder();
+        mappingLookup.getInferenceIdsForFields().entrySet().forEach(entry -> {
+            builder.put(entry.getKey(), new FieldInference(entry.getValue(), mappingLookup.sourcePaths(entry.getKey())));
+        });
+        fieldInferenceMap = builder.build();
+    }
+
+    public FieldInferenceMetadata(StreamInput in) throws IOException {
+        fieldInferenceMap = in.readImmutableOpenMap(StreamInput::readString, FieldInference::new);
+    }
+
+    public FieldInferenceMetadata(Map<String, FieldInference> fieldsToInferenceMap) {
+        fieldInferenceMap = ImmutableOpenMap.builder(fieldsToInferenceMap).build();
+    }
+
+    public boolean isEmpty() {
+        return fieldInferenceMap.isEmpty();
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeMap(fieldInferenceMap, (o, v) -> v.writeTo(o));
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.map(fieldInferenceMap);
+        return builder;
+    }
+
+    public static FieldInferenceMetadata fromXContent(XContentParser parser) throws IOException {
+        return new FieldInferenceMetadata(parser.map(HashMap::new, FieldInference::fromXContent));
+    }
+
+    public String getInferenceIdForField(String field) {
+        return getInferenceSafe(field, FieldInference::inferenceId);
+    }
+
+    private <T> T getInferenceSafe(String field, Function<FieldInference, T> fieldInferenceFunction) {
+        FieldInference fieldInference = fieldInferenceMap.get(field);
+        if (fieldInference == null) {
+            return null;
+        }
+        return fieldInferenceFunction.apply(fieldInference);
+    }
+
+    public Set<String> getSourceFields(String field) {
+        return getInferenceSafe(field, FieldInference::sourceFields);
```

Comment:
Let's expose the `fieldInferenceMap` directly, we don't need to dictate how the access should look like.

Reply:
Done 👍